src/performance.js

```diff
@@ -58,8 +66,48 @@ export class Performance {
 
     /** @const @private {!Array<TickEventDef>} */
     this.events_ = [];
+
+    /** @private {boolean} */
+    this.isViewerVisible_ = false;
+
+    this.win.addEventListener('unload', this.flush.bind(this));
   }
 
+  /**
+   * Gets primary amp core services.
+   * @return {!Performance}
+   */
+  init() {
+    this.viewer_ = viewerFor(this.win);
+    this.resources_ = resourcesFor(this.win);
+    return this;
+  }
+
+  /**
+   * Listens to viewer and resource events.
+   */
+  listen() {
+    if (!(this.viewer_ && this.resources_)) {
+      return;
+    }
+
+    // User requested document to be visible
+    this.viewer_.whenFirstVisible().then(() => {
+      this.isViewerVisible_ = true;
```

Comment:
I think there's a bug here. I'd think you need to initialize `isViewerVisible_` in the init method to the `viewer.isVisible()`. Generally, these kind of internal state variables are best to be moved to Viewer class I think. E.g. `isPrerendering()`, `hasBeenVisible()` in addition to the `isVisible` currently there. Why? Because if you do this here - you are running into all sort of risk conditions, such as "do we know for sure the document wasn't visible before?", etc.